### PR TITLE
Grammar tweaks

### PR DESCRIPTION
**Pull request by thec00n**
### Description

I am proposing the following two updates to the grammar:

-  ~~The loops `DoWhileStatement`, `WhileStatement` , `ForStatement` and the `IfStatement` contain `Statement` production rules to process statements. This is however not explicit enough even though `Statement` does contain `Block` along with the other statements. The production rule implies that an `IfStatement` for instance can produce another `IfStatement` after `'if' '(' Expression ')'`. This is not correct syntax because a nested `IfStatement` needs to be in a `Block` and can not have another `IfStatement` as an immediate parent.~~

-  A `VariableDeclaration` can not be optional when multiple `VariableDeclaration`s are defined as part of a `VariableDefinition`. 

This is more of question that came up when reading the grammar. Is there a specific reason why `Throw` is still a part of the grammar even though it is not a valid statement anymore since 0.5.x?

### CODE REVIEW

**Comment by axic** [comment]
>     * This is not correct syntax because a nested `IfStatement` needs to be in a `Block` and can not have another `IfStatement` as an immediate parent.

Are you sure? These seem to compile:
```
contract C {
  function f() pure public {
    if (true)
      if (true)
      {}
  }
}
```

```
contract C {
  function f(uint a, uint b) pure public {
    if (a % 2 == 1)
      if (a == 2)
      {}
  }
}
```

**Comment by thec00n** [comment]
Yeah agreed @axic. The grammar is correct in that respect. 

Looking at `Statement` again there seems to be cases though where the grammar does not produce valid Solidity code. Specifically the loop specific statements `Break` and `Continue` and `PlaceholderStatement` for modifiers are not valid within all statements. For example using `break` in the `if` statement is correct syntax according to the current grammar. 

```
contract C {
  function f() pure public {
    if (true) break;
  }
}
```

My attempt to write new production rules are in the latest updates to the PR. I am not very experienced with writing grammars so let me know what you guys think and if you generally agree that this is an issue in the current grammar.

File: docs/grammar.txt [hunk 1]
```diff
@@ -79,7 +79,7 @@ Break = 'break'
 Return = 'return' Expression?
 Throw = 'throw'
 EmitStatement = 'emit' FunctionCall
-VariableDefinition = (VariableDeclaration | '(' VariableDeclaration? (',' VariableDeclaration? )* ')' ) ( '=' Expression )?
+VariableDefinition = (VariableDeclaration | '(' VariableDeclaration (',' VariableDeclaration )* ')' ) ( '=' Expression )?
```
Comment (chriseth) [inline docs/grammar.txt]: Sorry, but this is not true. The following is valid code:
`(uint x, , uint y) = f()`.

File: docs/grammar.txt [hunk 2]
```diff
@@ -67,7 +67,7 @@ Statement = IfStatement | WhileStatement | ForStatement | Block | InlineAssembly
               Throw | EmitStatement | SimpleStatement ) ';'
 
 ExpressionStatement = Expression
-IfStatement = 'if' '(' Expression ')' Statement ( 'else' Statement )?
+IfStatement = 'if' '(' Expression ')' Block ( 'else' Block )?
```
Comment (chriseth) [inline docs/grammar.txt]: Again, `if (x) y = 8;` is valid code, no need for a block.

File: docs/grammar.txt [hunk 3]
```diff
@@ -21,12 +21,12 @@ UsingForDeclaration = 'using' Identifier 'for' ('*' | TypeName) ';'
 StructDefinition = 'struct' Identifier '{'
                      ( VariableDeclaration ';' (VariableDeclaration ';')* ) '}'
 
-ModifierDefinition = 'modifier' Identifier ParameterList? Block
+ModifierDefinition = 'modifier' Identifier ParameterList? BlockGeneralAndModifier
```
Comment (chriseth) [inline docs/grammar.txt]: While this change seems to properly restrict the grammar, I don't think it helps in understanding the language.

**Comment by chriseth** [comment]
A context-free grammar can never catch the full extend of the language. I would prefer to keep the grammar simple instead of trying to capture everything that a context-free grammar can possibly capture. Note that the parser also parses `break` statements outside loops. It is just a later stage in the compiler that performs these checks.

**Comment by chriseth** [comment]
Closing for now. Please feel free to re-open.

**Comment by thec00n** [comment]
> A context-free grammar can never catch the full extend of the language. I would prefer to keep the grammar simple instead of trying to capture everything that a context-free grammar can possibly capture. Note that the parser also parses `break` statements outside loops. It is just a later stage in the compiler that performs these checks.

@chriseth fair enough.

**Pull request closed without merging**